doc/obprobe.1
.Dd May 26, 2026
.Dt OBPROBE 1 URM
.Os "Open Babel" 3.2
.Sh NAME
.Nm obprobe
.Nd "create electrostatic probe grid"
.Sh SYNOPSIS
.Nm
.Op Ar OPTIONS
.Ar type
.Ar pchg
.Ar filename
.Sh DESCRIPTION
The
.Nm
tool creates a grid around a molecule, placing a probe
atom with a specified atom type and partial charge at each point to
calculate the MMFF94 energy.
This can be used for docking experiments to test hydrogen\-bond affinity,
electrostatic potential, etc.
Output is sent to standard output using the Gaussian Cube format.
.Sh OPTIONS
If no filename is given,
.Nm
will give all options including the example probes.
.Bl -tag -width flag
.It Fl s Ar stepsize
Set the resolution of the grid (stepsize)
.It Fl p Ar padding
Set the padding \(em extra distance on each side of the box formed by
the molecule.
.It Ar type
MMFF94 atom type
.It Ar pchg
MMFF94 partial charge
.El
.Sh EXAMPLES
Probe the file pyridines.sdf using a carbonyl oxygen \(em a hydrogen
bond acceptor with partial charge \-0.57:
.Dl "obprobe 7 \-0.57 pyridines.sdf"
.Pp
Probe the file pyridines.sdf using a phenyl carbon atom \(em a
hydrophobic atom with no partial charge:
.Dl "obprobe 37 0.0 pyridines.sdf"
.Sh SEE ALSO
.Xr obabel 1 .
.Pp
The web pages for Open Babel can be found at:
\%<\fBhttps://openbabel.org/\fR>
.Sh AUTHORS
The obprobe program was contributed by
.An Tim Vandermeersch .
.Pp
.An -nosplit
Open Babel is developed by a cast of many, including current maintainers
.An Geoff Hutchison ,
.An Chris Morley ,
.An Michael Banck ,
and innumerable others who have contributed fixes and additions.
For more contributors to Open Babel, see
\%<\fBhttps://openbabel.org/docs/Introduction/thanks.html\fR>
.Sh COPYRIGHT
Copyright (C) 2006\-2008 by Tim Vandermeersch
.br
Some portions Copyright (C) 2004\-2008 by Geoffrey R. Hutchison and
other contributors.
.Pp
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation version 2 of the License.
.Pp
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.


NEWS
Open Babel 2.5.0 (2017-XX-XX)

Open Babel 2.4.0 (2016-9-15)

Open Babel 2.3.0 (2010-10-23)

This release represents a major update and should be a stable upgrade,
strongly recommended for all users of Open Babel. Highlights include a
completely rewritten stereochemistry engine, Spectrophore fingerprint
generation, 2D depiction, improved 3D coordinate generation, conformer
searching, and more. Many formats are improved or added, including
CIF, PDBQT, SVG, and more. Improved developer API and scripting
support and many, many bug fixes are also included.

What's New from 2.2.3:

* Completely rewritten stereochemistry perception, including support
  for tetrahedral, square planar, and higher-order stereochemistry.
* Dramatically improved canonicalization algorithm (Note that in general, canonical SMILES have changed since the 2.2.x release.)
* 2D depiction, including SVG vector graphics generation using code from MCDL.
* New Spectrophore generation, contributed by Silicos NV.
* New ChargeMethod API including support for partial charge assignment
  from Gasteiger, MMFF94, QEq, QTPIE methods and plugin interface for
  adding more.
* Improved 3D coordinate generation.
* New conformer generation framework, including support for diverse
  conformer generation and genetic algorithm lowest-energy searching.
* Improved user documentation.
* Improved aromaticity / Kekule bond assignment.
* Improved unit test suite using the CMake-based CTest program.
* Improved support for crystallographic unit cells (e.g., in CIF format).
* Improved UFF force field method, including hypervalent 5, 6, 7 and higher coordination numbers.
* Support for the GAFF (Generalized Amber Force Field) method.
* Support for reading geometry optimizations as multiple conformers
  from Gaussian, GAMESS-US, and other quantum chemistry packages.
* Support for reading molecular orbital energies from quantum chemistry formats.
* Several memory leaks fixed.
* Fixed many compiler warnings.
* Fixed support for MinGW and Cygwin environments.
* Fixed bugs with Gaussian 09 output files.
* Latest released version of the InChI library (1.0.3) generating standard InChI.
* Many more bug fixes and small feature improvements.

==New Command-Line Operations==

* --canonical: Output atoms in canonical order for any file format (i.e., not just SMILES)
* --conformer: Run a conformer search on the input molecules (has many options)
* --gen2D: Generate a 2D depiction of the molecule
* --partialcharge <model>: Use the partial charge model supplied to
  generate charges (i.e., instead of default Gasteiger sigma model)
* --sort <descriptor>: Sort molecules by a specified descriptor
* --unique: Only output unique molecules (as determined by InChI generation)

==New File Formats==

Import & Export:
- DL-POLY CONFIG
- FHIaims XYZ
- PDBQT

Import only:
- DL-POLY HISTORY
- GULP output
- PWscf output
- Text

Export only:
- MNA (Multilevel Neighborhoods of Atoms)
- SVG vector graphics


Open Babel 2.2.3 (2009-07-31)

What's New from 2.2.2:

This release represents an important bug-fix upgrade, strongly recommended
for all users of Open Babel.

- Fixed bug in fingerprints in 2.2.2, where the default fingerprint
  format and bit length was changed inadvertently.
- Fixed detection of shared_ptr in tr1/memory.
- Fixed additional aromaticity / Kekule assignment bugs.
- Fixed several bugs in the MMCIF format.
- Additional bug fixes.

Open Babel 2.2.2 (2009-07-04)

What's New from 2.2.1:

This release represents a major bug-fix release and is a stable
upgrade, strongly recommended for all users of Open Babel. While there
may not be many new features, many crashes and other bugs have been
fixed since 2.2.1.

- Upgraded to the new InChI 1.02 release to produce standardized InChI
  and InChIKey output.
- Fixed many stereochemistry errors when reading/writing SMILES. This
  is part of a larger project which will be finished in the 2.3
  release.
- Fixed compilation and installation on Cygwin and MinGW platforms.
- Significantly improved aromaticity and Kekule bond assignment.
- Improved 2D -> 3D coordinate generation
- Improved coordinate generation using the --gen3d command-line
  operation
- Improved performance for coordinate generation.
- New --fillUC command-line operation for babel.
- Fixes to pH-dependent hydrogen addition.
- Added support for reading vibrational data from Molden, Molpro, and
  NWChem output files.
- Updated atomic radii from recent theoretical calculations.
- Fixed bug when reading gzip-compressed Mol2 or XML files.
- Close files after an error. Fixes a bug with Pybel where files would
  remain open.
- Many more bug fixes and small feature improvements.

==New File Formats==

Import & Export:
- Molpro input and output.
- VASP coordinate files (CONTCAR and POSCAR).

Open Babel 2.2.1 (2009-03-01)

What's New from 2.2.0:

This release represents a major bug-fix release and is a stable
upgrade, strongly recommended for all users of Open Babel. While there
may not be many new features, many crashes and other bugs have been
fixed since 2.2.0.

- Improved scripting interfaces, including Python 3 support and improved
  Java and C# support.
- Added support for MACCS fingerprints. Thanks to the RDKit project.
- Many fixes and enhancements to the force field code. In particular,
  the UFF force field implementation should handle many more molecules.
- Improved 3D coordinate generation, particularly with ring
  fragments. You can give this a try with the obgen utility.
- Fixed a variety of PDB import errors with atom types.
- Added support for reading charges and radii from PQR file formats.
- Added support for reading and writing unit cells in PDB formats.
- New "output" file format for taking generic ".out", ".log", and
  ".dat" files and reading with appropriate file type based on
  contents. Currently works extremely well for quantum chemistry
  packages.
- Added improved error handling and reporting when unable to load file
  formats.

tools/obprobe.cpp
/**********************************************************************
obprobe.cpp - This program will create a grid around a molecule, place
              a probe atom with specified type and partial charge at
              each grid point and calculate the energy. This energy is
              stored in the grid. MMFF94 only for now.

Copyright (C) 2006 Tim Vandermeersch
 
This file is part of the Open Babel project.
For more information, see <http://openbabel.org/>

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation version 2 of the License.
 
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
***********************************************************************/

// used to set import/export for Cygwin DLLs
#ifdef WIN32
#define USING_OBDLL
#endif

#include <openbabel/babelconfig.h>
#include <openbabel/base.h>
#include <openbabel/mol.h>
#include <openbabel/obiter.h>
#include <openbabel/griddata.h>
#include <openbabel/obconversion.h>
#include <openbabel/forcefield.h>
#ifndef _MSC_VER
  #include <unistd.h>
#endif
#include <cstdlib>

using namespace std;
using namespace OpenBabel;

int main(int argc,char **argv)
{
  char *program_name = argv[0];
  char *type;
  int c;
  double step, padding, pchg;
  string basename, filename = "", option;

  step    = 0.5;
  padding = 5.0;

  if (argc < 4) {
    cout << "Usage: obprobe [options] <type> <pchg> <filename>" << endl;
    cout << endl;
    cout << "  <type>: the probe MMFF94 atom type" << endl;
    cout << endl;
    cout << "  <pchg>: the probe's partial charge" << endl;
    cout << endl;
    cout << endl;
    cout << "Options:          Description:" << endl;
    cout << endl;
    cout << "  -s <stepsize>   step size" << endl;
    cout << endl;
    cout << "  -p <padding>    padding" << endl;
    cout << endl;
    cout << "Example probes:" << endl;
    cout << endl;
    cout << "<type>  <pchg>    Description:" << endl;
    cout << "  7     -0.57     carbonyl oxygen (HBA)" << endl;
    cout << " 21      0.4      hydroxyl hydrogen (HBD)" << endl;
    cout << " 37      0.0      phenyl carbon (hydrophobic)" << endl;
    exit(-1);
  } else {
    int ifile = 1;
    for (int i = 1; i < argc; i++) {
      option = argv[i];
      
      if ((option == "-s") && (argc > (i+1))) {
        string stepstr = argv[i+1];
        step = atof(stepstr.c_str());
        ifile += 2;
      }

      if ((option == "-p") && (argc > (i+1))) {
        string paddingstr = argv[i+1];
        padding = atof(paddingstr.c_str());
        ifile += 2;
      }
    }
    
    type = argv[ifile];
    pchg = atof(argv[ifile+1]);
    basename = filename = argv[ifile+2];
    size_t extPos = filename.rfind('.');

    if (extPos!= string::npos) {
      basename = filename.substr(0, extPos);
    }


  }

  // Find Input filetype
  OBConversion conv;
  OBFormat *format_in = conv.FormatFromExt(filename.c_str());
  OBFormat *format_out = conv.FormatFromExt(".cube");
    
  if (!format_in || !format_out || !conv.SetInAndOutFormats(format_in, format_out)) {
    cerr << program_name << ": cannot read input/output format!" << endl;
    exit (-1);
  }

  ifstream ifs;

  // Read the file
  ifs.open(filename.c_str());
  if (!ifs) {
    cerr << program_name << ": cannot read input file!" << endl;
    exit (-1);
  }

  OBForceField* pFF = OBForceField::FindForceField("MMFF94");
  if (!pFF) {
    cerr << program_name << ": could not find forcefield 'MMFF94'." <<endl;
    exit (-1);
  }
  pFF->SetLogFile(&cerr);
  pFF->SetLogLevel(OBFF_LOGLVL_NONE);

  OBMol mol;
  char buffer[BUFF_SIZE];
  for (c=1;;c++) {
    mol.Clear();
    if (!conv.Read(&mol, &ifs))
      break;
    if (mol.Empty())
      break;

    if (!pFF->Setup(mol)) {
      cerr << program_name << ": could not setup force field." << endl;
      exit (-1);
    }
    
    OBGridData* gd = pFF->GetGrid(step, padding, type, pchg);
    mol.SetData(gd);

    ofstream ofs;
    snprintf(buffer, BUFF_SIZE, "%s_%s_%f.cube", basename.c_str(), type, pchg);
    ofs.open(buffer);
    if (!ofs) {
      cerr << program_name << ": cannot read input file!" << endl;
      exit (-1);
    }
    if (!conv.Write(&mol, &ofs)) {
      cerr << program_name << ": could not setup force field." << endl;
    }

    ofs.close();

  } // end for loop

  return(0);
}
